src/plugins/spec/testmod_spec.c

```diff
@@ -608,15 +624,16 @@ int main (int argc, char ** argv)
 
 	init (argc, argv);
 
-	test_default ();
-	test_assign_condition ();
-	test_wildcard ();
-	test_require ();
-	test_logMissing ();
-	test_array ();
-	test_require_array ();
-	test_array_member ();
-	// test_remove_meta ();
+	test_hook_copy_default (true);
```

Comment:
Should we test all of them for both the get and set operations? Should there be differences between the operations?
For now, e.g. the set operation does not create default:/ keys.

Reply:
The copy should behave the same in `kdbGet` and `kdbSet` that's why I used the same name.


That said, the current set from `spec` is a bit broken.

To make validation for newly added keys during `kdbSet` work properly, we need to copy `meta:/` keys from `spec:/` to other namespaces at the start of `kdbSet` (before `prestorage`). This should behave the same as in `kdbGet` and the new `meta:/` keys should be visible outside `kdbSet` I think (see https://github.com/ElektraInitiative/libelektra/pull/4484#discussion_r979454050).

However, we don't want to store the copied `meta:/` keys in the other namespace, so we need to remove them before the `storage` phase. To do this we should simply check for every metakey of every key in a namespace other than `spec:/` whether it (the metakey) is pointer-equal to the corresponding metakey on the equivalent `spec:/` key. So something like:

```c
KeySet * specMeta = keyMeta (specKey);
KeySet * meta = keyMeta (key);

for (elektraCursor it = 0; i < ksGetSize (specMeta); i++)
{
    Key * m = ksAtCursor (specMeta, it);
    if (ksLookup (meta, m, 0) == m)
    {
        keySetMeta (key, keyName (m), NULL);
    }
}
```

Reply:
The differences between a set and get operation for the spec plugin were copied from the existing implementation:

- There can be different settings for the conflict handling 
  - This can probably be removed
- neither `assign/condition` nor `default` are processed in set 
  - We can probably also execute this on set
- set operation sets `internal/spec/remove` on `array`
  - remove or also do it in get?
- set removes `internal/spec/array/validated`
  - remove or also do it in get?
- set does this thing here, I'm not sure what it is
    - ???
```c
if (!isKdbGet)
{
	keySetMeta (specKey, "internal/spec/array/validated", NULL);

	if (keyGetMeta (specKey, "internal/spec/array") == NULL && keyGetMeta (specKey, "internal/spec/remove") == NULL)
	{
		ksAppendKey (returned, specKey);
	}
}
```

Reply:
> There can be different settings for the conflict handling
> * This can probably be removed

Yes, all the conflict handling settings should be removed entirely. During `get` all conflicts should be warnings during `set` they should be errors (unless @markus2330 thinks some conflicts should never be errors). That part could be handled from the `libelektra-kdb` side with the "error blocking trick", so the `spec` would be the same.


> neither `assign/condition` nor `default` are processed in set

Not sure about `assign/condition` and whether we actually want to keep that (@markus2330 ?), but `default` should definitely be processed in `set`. See also https://github.com/ElektraInitiative/libelektra/pull/4484#discussion_r979454050



> * set operation sets `internal/spec/remove` on `array`
> * set removes `internal/spec/array/validated`
> * set does this thing here, I'm not sure what it is

All of the stuff related to array handling is very complicated and weird. Part of the weirdness comes from the fact that on `master` the `spec` plugin is only called once during `kdbSet` and somehow there should still be validation during `kdbSet`, but also `spec` should cleanup the copied data (obviously not possible).

See also #2700 and #4061

---

However, please don't waste time on this. Like I said already: Just try to imitate what `spec` does on `master`. The existing tests obviously also expect this behaviour. Once that is done and merged, you can create another PR with the actual updates for `spec` (probably after `new-backend` is merged in `master`).

Reply:
Especially, the conflict handling should remain the same as on `master` for now. Changing this will probably break some tests and take more time than might expect.

Reply:
> However, please don't waste time on this. [...] Once that is done and merged, you can create another PR with the actual updates for spec.

I agree, this is probably the correct method, so I can continue with the other hooks. 

Please take a look at my comment below with the test cases. I think those still need to be resolved before we can merge it.

Reply:
It looks like we got someone who will work on the spec plugin! :rocket:

So we should soon have a coordination meeting about spec-related topics.

Reply:
> Not sure about assign/condition and whether we actually want to keep that (@markus2330 ?)

No, we can remove it. Defaults are a superior way. It should be possible via `namespace` spec to give `default` higher priority than others, giving something like an assignment via defaults.